opal/mca/btl/iwarp/btl_iwarp_async.c

```diff
@@ -0,0 +1,443 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2008-2009 Mellanox Technologies. All rights reserved.
+ * Copyright (c) 2007-2013 Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2006-2007 Voltaire All rights reserved.
+ * Copyright (c) 2009-2010 Oracle and/or its affiliates.  All rights reserved
+ * Copyright (c) 2013-2018 Los Alamos National Security, LLC. All rights
+ *                         reserved.
+ * Copyright (c) 2014      Intel, Inc. All rights reserved.
+ * Copyright (c) 2014      Bull SAS.  All rights reserved.
+ * Copyright (c) 2015      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+
+#include "opal_config.h"
+
+#include <infiniband/verbs.h>
+#include <fcntl.h>
+#include <sys/poll.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include "opal/util/show_help.h"
+#include "opal/util/proc.h"
+
+#include "opal/mca/btl/base/base.h"
+#include "btl_iwarp.h"
+#include "btl_iwarp_mca.h"
+#include "btl_iwarp_async.h"
+#include "btl_iwarp_proc.h"
+#include "btl_iwarp_endpoint.h"
+
+static opal_list_t ignore_qp_err_list;
+static opal_mutex_t ignore_qp_err_list_lock;
+static int32_t btl_iwarp_async_device_count = 0;
+
+struct mca_btl_iwarp_async_poll {
+    int active_poll_size;
+    int poll_size;
+    struct pollfd *async_pollfd;
+};
+typedef struct mca_btl_iwarp_async_poll mca_btl_iwarp_async_poll;
+
+typedef struct {
+    opal_list_item_t super;
+    struct ibv_qp *qp;
+} mca_btl_iwarp_qp_list;
+
+OBJ_CLASS_INSTANCE(mca_btl_iwarp_qp_list, opal_list_item_t, NULL, NULL);
+
+static const char *iwarp_event_to_str (enum ibv_event_type event);
+
+/* Function converts event to string (name)
+ * Open Fabris don't have function that do this job :(
+ */
+static const char *iwarp_event_to_str (enum ibv_event_type event)
+{
+    switch (event) {
+    case IBV_EVENT_CQ_ERR:
+        return "IBV_EVENT_CQ_ERR";
+    case IBV_EVENT_QP_FATAL:
+        return "IBV_EVENT_QP_FATAL";
+    case IBV_EVENT_QP_REQ_ERR:
+        return "IBV_EVENT_QP_REQ_ERR";
+    case IBV_EVENT_QP_ACCESS_ERR:
+        return "IBV_EVENT_QP_ACCESS_ERR";
+    case IBV_EVENT_PATH_MIG:
```

Comment:
Might want to check what events will actually be reported here.  E.g., if there's no APM in iWARP, then you'll never get path migration events.